src/BinOp.chpl

```diff
@@ -1425,15 +1427,21 @@ module BinOp
        (val.type == bigint && (l.etype == int || l.etype == uint || l.etype == bool)) {
       select op {
         when "+" {
-          tmp = la + val;
+          forall (t, li) in zip(tmp, la) with (var local_val = val) {
+            t = li + local_val;
+          }
```

Comment:
@ronawho will this result in any extra overhead in the `val.type != bigint` cases? I was thinking it probably wouldn't be much if any since `int/uint` would be trivially copyable. If I'm wrong, do you think updating to something like the below would be a good idea
```chapel
if val.type != bigint {
  tmp = la + val;
}
else {
  forall (t, li) in zip(tmp, la) with (var local_val = val) {
    t = li + local_val;
  }
}
```

Reply:
I have another question, I have two versions of the same chunk of code
```chapel
if has_max_bits {
  forall t in tmp with (var local_val = val, var local_max_size = max_size) {
    t *= local_val;
    t &= local_max_size;
  }
}
else {
  forall t in tmp with (var local_val = val) {
    t *= local_val;
  }
}
```
and
```chapel
forall t in tmp with (var local_val = val, var local_max_size = max_size) {
  t *= local_val;
  if has_max_bits {
    t &= local_max_size;
  }
}
```
I like the second one better visually, but I felt like the first one would be faster (just checking the condition once instead of every iteration). However the performance seems to be the same locally. Maybe it's just negligible compared to the bottle neck of the bigint operation?

Reply:
```chpl
forall (t, li) in zip(tmp, la) with (var local_val = val) {
  t = li + local_val;
}
```

Shouldn't have any overhead compared to:

```chpl
tmp = la + val;
```

for plain old numeric types. 


The promoted version is equivalent to (and transformed by the compiler into):

```chpl
forall (t, li) in zip(tmp, la) {
  t = li + val;
}
```

In this case the compiler will also detect that `val` is const and trivially-copyable, so it'll "remote value forward" it, which is basically means copy it's value into the argument bundled that's used to implement on-statements under the covers and avoid having to issue a `GET` for it. By introducing the task local var, there's an extra local copy at task startup, but this is so very trivial that you'll never notice the overhead.

Reply:
```chpl
forall t in tmp with (var local_val = val, var local_max_size = max_size) {
  t *= local_val;
  if has_max_bits {
    t &= local_max_size;
  }
}
```

This version will have some overhead, but probably a fairly small amount. It's adding a branch to every iteration, but the branch will be correctly predicted since it doesn't change over the course of the loop. So that effectively means just a few extra memory reads each iteration, which should be fast, but so are the other operations in the loop (at least for small bigints).


Here's a perf test to check:

```chpl
use TestBase, BigInteger;

config const size=10**8;
config const has_max_bits = false;

proc testit(type elemType) {
  writeln("Testing '", elemType:string, "'");
  const D = makeDistDom(size*numLocales);
  var tmp: [D] elemType;
  var val = 5: elemType;
  var max_size = 64: elemType;

  var d: Diags;

  d.start();
  if has_max_bits {
    forall t in tmp with (var local_val = val, var local_max_size = max_size) {
      t *= local_val;
      t &= local_max_size;
    }
  } else {
    forall t in tmp with (var local_val = val) {
      t *= local_val;
    }
  }
  d.stop("Outer Branch");

  d.start();
  forall t in tmp with (var local_val = val, var local_max_size = max_size) {
    t *= local_val;
    if has_max_bits {
      t &= local_max_size;
    }
  }
  d.stop("Inner Branch");
}

testit(int); writeln();
testit(bigint);
```


```
16-node-cs-hdr: arkouda (master) $ make test-bin/BigintOpSpeedTest
16-node-cs-hdr: arkouda (master) $ ./test-bin/BigintOpSpeedTest -nl 16 --size=$((10**8))
Testing 'int(64)'
Outer Branch: 0.01s
Inner Branch: 0.01s

Testing 'bigint'
Outer Branch: 0.05s
Inner Branch: 0.06s

16-node-cs-hdr: arkouda (master) $ ./test-bin/BigintOpSpeedTest -nl 16 --size=$((2**30))
Testing 'int(64)'
Outer Branch: 0.11s
Inner Branch: 0.12s

Testing 'bigint'
Outer Branch: 0.61s
Inner Branch: 0.64s
```

So there is a minor perf hit from the inner branch, but I'd probably still go with that to avoid code duplication (increased compilation times and more code to maintain)